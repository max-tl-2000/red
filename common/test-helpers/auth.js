/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import moment from 'moment'; // eslint-disable-line red/no-moment
import { expect } from 'chai';

export const verifyTokenExpiration = (decodedToken, expectedTokenExpiration) => {
  const iat = moment.unix(decodedToken.iat);
  const exp = moment.unix(decodedToken.exp);

  const result = exp.diff(iat, expectedTokenExpiration.substring(expectedTokenExpiration.length - 1));
  const expectedValue = expectedTokenExpiration.substring(0, expectedTokenExpiration.length - 1);
  expect(result).to.equal(+expectedValue);
};
