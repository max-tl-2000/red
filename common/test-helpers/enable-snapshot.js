/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import chai from 'chai';
import chaiJestSnapshot from 'chai-jest-snapshot';

const enableSnapshot = () => {
  if (!global.before) return;
  chai.use(chaiJestSnapshot);
  before(() => {
    chaiJestSnapshot.resetSnapshotRegistry();
  });

  beforeEach(function beforeEach() {
    const { currentTest } = this;
    chaiJestSnapshot.setFilename(`${currentTest.file}.chai-snap`);
    chaiJestSnapshot.setTestName(currentTest.fullTitle());
  });
};

enableSnapshot();

export { editUnpredictable } from '../helpers/snapshot-helper';
