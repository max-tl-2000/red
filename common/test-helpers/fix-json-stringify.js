/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

const stringify = JSON.stringify;

before(() => {
  JSON.stringify = obj => {
    const seen = [];

    return stringify(obj, (key, val) => {
      if (typeof val === 'object') {
        if (seen.indexOf(val) >= 0) {
          return;
        }
        seen.push(val);
      }
      return val; // eslint-disable-line
    });
  };
});

after(() => {
  JSON.stringify = stringify;
});
