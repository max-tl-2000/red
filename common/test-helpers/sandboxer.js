/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

let _sinon;
let sandboxes = [];

function create() {
  // this is for when this module is used in karma
  _sinon = _sinon || global.sinon;
  const sandbox = _sinon.sandbox.create();
  sandboxes.push(sandbox);

  return sandbox;
}

function restore() {
  sandboxes.forEach(sandbox => sandbox.restore());
  sandboxes = [];
}

function setSinon(sinon) {
  _sinon = sinon;
}

export { create, restore, setSinon };
