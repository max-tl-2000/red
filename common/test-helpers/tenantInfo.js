/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export const TEST_TENANT_ID = 'af20764d-1e0d-41a5-be54-3766215f89a4';
