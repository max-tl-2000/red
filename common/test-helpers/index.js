/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import { unmountComponentAtNode } from 'react-dom';
import when from './when';
import and from './and';
import * as overrider from './overrider';
import * as sandboxer from './sandboxer';
import sleep from '../helpers/sleep';

sandboxer.setSinon(sinon);

const expect = chai.expect;
const assert = chai.assert;

chai.use(sinonChai);

let div;

if (global.document) {
  beforeEach(() => {
    div = document.createElement('div');
    div.setAttribute('id', 'sandbox');

    document.body.appendChild(div);
  });
}

afterEach(() => {
  if (global.document && div) {
    unmountComponentAtNode(div);
    document.body.removeChild(div);
    div = null;
  }

  overrider.restore();
  sandboxer.restore();
});

const getSandBoxDiv = () => div;

// helper method to trigger a real click event in the browser
// jquery.fn.click is not the same as this as the event need
// to also generate the capture phase
const triggerMouseEvent = (ele, evt) => {
  const mEvent = document.createEvent('MouseEvent');
  // interface
  // event.initMouseEvent(type, canBubble, cancelable, view,
  //                    detail, screenX, screenY, clientX, clientY,
  //                    ctrlKey, altKey, shiftKey, metaKey,
  //                    button, relatedTarget);
  mEvent.initMouseEvent(evt, true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);

  ele.dispatchEvent(mEvent);
};

export { getSandBoxDiv, triggerMouseEvent, overrider, expect, assert, sinon, when, sandboxer, and, sleep };
