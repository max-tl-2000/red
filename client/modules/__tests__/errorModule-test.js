/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

/* eslint-disable global-require */
import { override, restore } from '../../../common/test-helpers/overrider';

const { mockModules } = require('../../../common/test-helpers/mocker').default(jest);

describe('errorModule', () => {
  let windowMock;
  let consoleMock;
  let logger;

  beforeEach(() => {
    windowMock = {
      __events: {},
      navigator: { userAgent: 'Mock User Agent' },
      location: { href: 'https://mock/location/to/resource' },
      addEventListener: jest.fn((evt, handler) => (windowMock.__events[evt] = handler)),
    };

    logger = { error: jest.fn(), warn: jest.fn() };
    consoleMock = { log: jest.fn(), error: jest.fn(), warn: jest.fn() };

    override(console, consoleMock);

    mockModules({
      '../../../common/helpers/globals': {
        window: windowMock,
      },
    });
  });

  afterEach(() => {
    jest.resetModules();
    restore();
  });

  it('should throw if no apiClient is provided', () => {
    const { initErrorModule } = require('../errorModule');

    const fn = () => initErrorModule();

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should throw if no id is provided', () => {
    const { initErrorModule } = require('../errorModule');

    const fn = () => initErrorModule(logger, {});

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  describe('skipping known errors', () => {
    it('should skip logging errors if they are prevented by CORS policy', () => {
      const { initErrorModule } = require('../errorModule');

      initErrorModule(logger, { id: 'mockId' });

      windowMock.onerror('Script error.', '', 0, 0, null);

      expect(consoleMock.log).toHaveBeenCalledWith('Ignored error', 'Script error.', {});

      expect(logger.error).not.toHaveBeenCalled();
    });

    it('should skip logging errors if it match the "ResizeObserver loop limit exceeded"', () => {
      const { initErrorModule } = require('../errorModule');

      initErrorModule(logger, { id: 'mockId' });

      windowMock.onerror('ResizeObserver loop limit exceeded', '', 0, 0, null);

      expect(consoleMock.warn).toHaveBeenCalledWith('Ignored error', 'ResizeObserver loop limit exceeded', {});

      expect(logger.error).not.toHaveBeenCalled();
    });

    it('should skip logging errors if it match the "Blocked a frame with origin"', () => {
      const { initErrorModule } = require('../errorModule');

      initErrorModule(logger, { id: 'mockId' });

      windowMock.onerror('Blocked a frame with origin', '', 0, 0, null);

      expect(consoleMock.warn).toHaveBeenCalledWith('Ignored error', 'Blocked a frame with origin', {});

      expect(logger.error).not.toHaveBeenCalled();
    });
  });

  it('should call logger.error if an error is received with a proper Error object', () => {
    const { initErrorModule } = require('../errorModule');

    initErrorModule(logger, { id: 'mockId' });
    const mockError = {
      message: 'This is a mock error',
      stack: `
        This is a mock stack at file:10:1
        This is another mock at file:20:1
      `,
    };

    windowMock.onerror('An interesting error', 'some/path/to/file', 17, 20, mockError);

    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('should call the old onerror handler if one was defined', () => {
    const { initErrorModule } = require('../errorModule');
    const oldHandler = jest.fn();

    windowMock.onerror = oldHandler;

    initErrorModule(logger, { id: 'mockId' });

    windowMock.onerror('An interesting error', 'some/path/to/file', 17, 20, null);

    expect(oldHandler).toHaveBeenCalled();
  });

  it('should add a listener to `unhandledrejection` event', async () => {
    const { initErrorModule } = require('../errorModule');

    initErrorModule(logger, { id: 'mockId' });

    windowMock.__events.unhandledrejection({
      promise: {},
      reason: 'error',
    });

    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('in production unhandled rejections should not be printed to console', async () => {
    override(process.env, { NODE_ENV: 'production' });
    // this is achieved by calling e.preventDefault(); so in this case we will just check
    // that preventDefault was called
    const { initErrorModule } = require('../errorModule');

    initErrorModule(logger, { id: 'mockId' });
    const args = { promise: {}, reason: 'error', preventDefault: jest.fn() };
    windowMock.__events.unhandledrejection(args);

    expect(args.preventDefault).toHaveBeenCalled();
  });
});
