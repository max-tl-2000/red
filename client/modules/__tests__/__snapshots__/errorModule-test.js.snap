// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorModule should add a listener to \`unhandledrejection\` event 1`] = `
Array [
  Array [
    Object {
      "reason": "error",
      "stack": undefined,
    },
    "mockId, unhandled rejection",
  ],
]
`;

exports[`errorModule should call logger.error if an error is received with a proper Error object 1`] = `
Array [
  Array [
    Object {
      "col": 20,
      "errorMessage": "An interesting error",
      "errorObject": Object {
        "message": "This is a mock error",
        "stack": "
        This is a mock stack at file:10:1
        This is another mock at file:20:1
      ",
      },
      "file": "some/path/to/file",
      "line": 17,
    },
    "mockId, application error",
  ],
]
`;

exports[`errorModule should throw if no apiClient is provided 1`] = `"logger object not provided"`;

exports[`errorModule should throw if no id is provided 1`] = `"id not provided"`;
